doc/guides/smf/index.rst

```diff
@@ -0,0 +1,269 @@
+framework's functions casts the user defined object to the ``struct smf_ctx``
+type with the following macro: ``SMF_CTX(o)``
+
+For example instead of doing this ``(struct smf_ctx *)&user_obj``, you could
+use ``SMF_CTX(&user_obj)``.
+
+The following macros can be used for easy state creation:
+
+* :c:macro:`HSM_CREATE_STATE` Create a hierachical state
+* :c:macro:`FSM_CREATE_STATE` Create a flat state
+
+To enable a flat state machine, define the :kconfig:`CONFIG_SMF_FLAT` option.
+
+State Machine Creation
+======================
+
+A state machine is created by defining a table of states that's indexed by an
+enum. For example::
+
+   enum demo_state { S0, S1, S2 };
+
+   const struct smf_state demo_states {
+      [S0] = FSM_CREATE_STATE(s0_entry, s0_run, s0_exit),
+      [S1] = FSM_CREATE_STATE(s1_entry, s1_run, s1_exit),
+      [S2] = FSM_CREATE_STATE(s2_entry, s2_run, s2_exit)
+   };
+
+To set the initial state, the ``smf_set_initial`` function should be
+called. It has the following prototype:
+``void smf_set_initial(smf_ctx *ctx, smf_state *state)``
+
+To transition from one state to another, the ``smf_set_state`` function is
+used and it has the following prototype:
+``void smf_set_state(smf_ctx *ctx, smf_state *state)``
+
+**NOTE:** While the state machine is running, smf_set_state should only be
+called from the Entry and Run functions. Calling smf_set_state from the Exit
+functions doesn't make sense and will generate a warning.
+
+State Machine Execution
+=======================
+
+To run the state machine, the ``smf_run_state`` function should be called in
+some application dependent way. An application should cease calling
+smf_run_state if it returns a non-zero value. The function has the following
+prototype: ``int32_t smf_run_state(smf_ctx *ctx)``
+
+State Machine Termination
+=========================
+
+To terminate the state machine, the ``smf_terminate`` function should be
+called. It can be called from the entry, run, or exit action. The function
+takes a non-zero user defined value that's returned by the ``smf_run_state``
+function. The function has the following prototype:
+``int32_t smf_terminate(smf_ctx *ctx, int32_t val)``
+
+Flat State Machine Example
+==========================
+
+This example turns the following state diagram into code using the SMF, where
+the initial state is S0.
+
+.. figure:: img/flat.svg
+    :align: center
+    :alt: Flat state machine diagram
+
+Code::
+
+	#include <smf.h>
+
+	/* Forward declaration of state table */
+	static const struct smf_state demo_states[];
+
+	/* List of demo states */
+	enum demo_state { S0, S1, S2 };
+
+	/* User defined object */
+	struct s_object {
+		/* This must be first */
+		struct smf_ctx ctx;
+
+		/* Other state specific data add here */
+	} s_obj;
+
+	/* State S0 */
+	static void s0_entry(void *o)
+	{
+		/* Do something */
+	}
+	static void s0_run(void *o)
+	{
+		smf_set_state(SMF_CTX(s_obj), &demo_states[S1]);
```

Comment:
@sambhurst I think & is needed, same as in other SMF_CTX calls

Reply:
Done